File: backend/app/schemas/cv.py
"""CV Studio schemas: source-document extraction + CV records.

`CvSourceExtraction` is the validated shape stored on `documents.extraction`
for kind="cv" files. `OcrPagesResult` is the structured AI output for the
vision-OCR fallback. CV CRUD schemas cover the living
record; versioning snapshots land with's endpoints.
"""

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator

from app.models.enums import CvKind, CvPageSize, CvStatus


class CvSourcePage(BaseModel):
    """One source page: recovered text + where it came from."""

    index: int = Field(ge=0)
    text: str = ""
    source: Literal["text_layer", "ocr_vision", "ocr_tesseract"]
    confidence: float = Field(default=1.0, ge=0.0, le=1.0)
    image_path: Optional[str] = None


class CvSourceExtraction(BaseModel):
    """Validated `documents.extraction` payload for kind="cv" files."""

    pages: list[CvSourcePage] = Field(default_factory=list, max_length=50)
    full_text: str = ""
    has_text_layer: bool = False
    ocr_used: bool = False
    content_sha256: str = Field(default="", max_length=64)
    engine: dict = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class OcrPage(BaseModel):
    """One OCR'd page as returned by the vision model."""

    index: int = Field(ge=0)
    text: str = Field(min_length=1)
    confidence: float = Field(default=0.8, ge=0.0, le=1.0)


class OcrPagesResult(BaseModel):
    """Structured output of the vision OCR task (one call per page batch)."""

    pages: list[OcrPage] = Field(max_length=10)


class CvContextRef(BaseModel):
    """A typed context item reference."""

    source_key: str = Field(min_length=1, max_length=60)
    item_id: str = Field(min_length=1, max_length=64)


class CvContextSelection(BaseModel):
    """Per-CV context selection: include-all-minus / none-plus / custom."""

    mode: Literal["all", "none", "custom"] = "all"
    include: list[CvContextRef] = Field(default_factory=list, max_length=500)
    exclude: list[CvContextRef] = Field(default_factory=list, max_length=500)


class CvDocumentCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    kind: CvKind = CvKind.RESUME
    language: str = Field(default="en", min_length=2, max_length=10)
    page_size: CvPageSize = CvPageSize.A4
    max_pages: int = Field(default=1, ge=1, le=10)
    target_posting_id: Optional[uuid.UUID] = None
    template_id: Optional[uuid.UUID] = None
    photo_document_id: Optional[uuid.UUID] = None
    source_document_id: Optional[uuid.UUID] = None
    context: CvContextSelection = Field(default_factory=CvContextSelection)


class CvDocumentUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    status: Optional[CvStatus] = None
    language: Optional[str] = Field(default=None, min_length=2, max_length=10)
    page_size: Optional[CvPageSize] = None
    max_pages: Optional[int] = Field(default=None, ge=1, le=10)
    target_posting_id: Optional[uuid.UUID] = None
    template_id: Optional[uuid.UUID] = None
    photo_document_id: Optional[uuid.UUID] = None
    working_content: Optional[dict] = None
    context: Optional[CvContextSelection] = None


class CvVersionOut(BaseModel):
    id: uuid.UUID
    version: int
    content: dict
    context_resolution: dict
    content_hash: str
    created_by: str
    created_at: datetime

    model_config = {"from_attributes": True}


class CvDocumentOut(BaseModel):
    id: uuid.UUID
    title: str
    kind: CvKind
    target_posting_id: Optional[uuid.UUID] = None
    template_id: Optional[uuid.UUID] = None
    photo_document_id: Optional[uuid.UUID] = None
    language: str
    page_size: CvPageSize
    max_pages: int
    status: CvStatus
    working_content: dict
    context: dict
    source_document_id: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime
    latest_version: Optional[int] = None

    model_config = {"from_attributes": True}

    @field_validator("context", mode="before")
    @classmethod
    def _validate_context(cls, value):  # noqa: ANN001 - raw JSONB
        if isinstance(value, dict) and value:
            CvContextSelection.model_validate(value)
        return value


class CvContextItemInfo(BaseModel):
    """One resolvable item as shown in the editor's source tree."""

    item_id: str
    label: str
    detail: str = ""


class CvContextSourceOut(BaseModel):
    key: str
    label: str
    description: str
    items: list[CvContextItemInfo] = Field(default_factory=list)


class CvContextSourcesOut(BaseModel):
    sources: list[CvContextSourceOut]


class CvContextItemRef(BaseModel):
    """Trace ref: one resolved item inside a compiled snapshot."""

    source_key: str
    item_id: str
    label: str = ""
    updated_at: Optional[str] = None


class CvResolutionOut(BaseModel):
    snapshot: dict
    snapshot_index: dict[str, list[str]]
    items: list[CvContextItemRef]
    resolved_at: datetime


class CvPreviewOut(BaseModel):
    html: str
    metrics: dict
    resolution: CvResolutionOut
    blocks: list[dict] = Field(default_factory=list)


class CvCompileOut(BaseModel):
    version: CvVersionOut
    html: str
    metrics: dict


class CvContextStatusOut(BaseModel):
    has_baseline: bool
    stale: bool
    changed: list[CvContextItemRef] = Field(default_factory=list)
    added: list[CvContextItemRef] = Field(default_factory=list)
    removed: list[CvContextItemRef] = Field(default_factory=list)

File: frontend/src/types/cv.ts
export interface CvContextRef {
  source_key: string;
  item_id: string;
}

export interface CvContextSelection {
  mode: "all" | "none" | "custom";
  include: CvContextRef[];
  exclude: CvContextRef[];
  /** Plan 62: prefer applies matching synthesized variants at resolve time. */
  synth_mode?: "off" | "prefer";
  /** Plan 72 follow-up: per-item variant pins, `{"source_key:item_id": synth_id}`. */
  synth_pins?: Record<string, string>;
}

export type CvSynthAction =
  | "summarize"
  | "detail"
  | "restyle"
  | "posting_fit"
  | "translate";

export interface CvSynthPayload {
  description?: string;
  summary?: string;
  bullets?: string[];
}

export interface CvSynthVoice {
  language: string;
  tone?: string;
  length?: string;
  action?: string;
}

export interface CvSynthItem {
  id: string;
  scope: "item" | "summary";
  variant_key: string;
  target_posting_id?: string | null;
  source_refs: CvContextRef[];
  source_state: { source_key: string; item_id: string; content_hash?: string | null }[];
  payload: CvSynthPayload;
  voice: CvSynthVoice;
  status: "draft" | "active" | "archived";
  source: "ai" | "manual";
  verified: boolean;
  stale: boolean;
  orphaned: boolean;
  last_used_at?: string | null;
  created_at?: string;
  verdict?: string;
}

export interface CvSynthGenerateOut {
  job_id?: string | null;
  items: CvSynthItem[];
}

export interface CvSynthGenerateRequest {
  refs: CvContextRef[];
  action: CvSynthAction;
  scope?: "item" | "summary";
  posting_id?: string;
  language?: string;
  target_language?: string;
  tone?: string;
  length?: string;
  variant_key?: string;
  translate_of?: string;
  regenerate_of?: string;
}

export type CvGenerateSectionKind =
  | "summary"
  | "experience"
  | "projects"
  | "volunteer"
  | "education"
  | "certifications"
  | "skills"
  | "languages"
  | "achievements"
  | "interests";

export interface CvGenerateRequest {
  target_posting_id?: string | null;
  /** Pasted job-posting body — used when no saved posting is targeted
   * (backend caps at 5000 chars; first line names the CV title). */
  posting_text?: string;
  /** "ai" = the background run picks the best template (the modal can
   * submit instantly); "none" = studio default or an explicit id. */
  template_pick?: "none" | "ai";
  language?: string;
  tone?: "professional" | "warm" | "concise" | "confident" | null;
  length?: "concise" | "standard" | "detailed";
  max_pages?: number;
  template_id?: string | null;
  include_photo?: boolean;
  sections?: CvGenerateSectionKind[];
  context?: CvContextSelection;
  notes?: string;
}

export interface CvGenerateResult {
  cv_id: string;
  title: string;
  version: number;
  lint: Record<string, unknown>;
  warnings: string[];
  fallback_sections: string[];
  plan_fallback: boolean;
  synth_applied?: Record<string, string>;
  synth_proposed?: {
    source_key: string;
    item_id: string;
    action: string;
    synth_item_id: string;
  }[];
  polish?: CvPolishTrace;
}

export interface CvSynthPreviewItem {
  source_key: string;
  item_id: string;
  synth_id: string;
  variant_key: string;
  stale: boolean;
}

export interface CvSynthPreview {
  items: CvSynthPreviewItem[];
  total: number;
}

export interface CvPolishTrace {
  request?: {
    notes?: string;
    target_posting_id?: string;
    tone?: string;
    length?: string;
    language?: string;
    max_pages?: number;
    sections?: string[];
    include_photo?: boolean;
  };
  run?: {
    started_at?: string;
    finished_at?: string;
    resumed_from?: { job_id?: string };
    stages?: { node: string; at: string; note?: string }[];
  };
  outcome?: { status: string; error?: string };
  llm_calls?: {
    id?: string;
    task: string;
    stage?: string | null;
    status?: string;
    provider?: string | null;
    model?: string | null;
    prompt_version?: string | null;
    tokens_in?: number | null;
    tokens_out?: number | null;
    latency_ms?: number | null;
  }[];
  iterations?: {
    n: number;
    started_at?: string;
    finished_at?: string;
    summary?: string;
    issues?: {
      level: "fail" | "warn" | "info";
      area: string;
      message: string;
    }[];
    ops?: { op: string; ok: boolean; detail?: string }[];
    coverage?: {
      included?: { source_key: string; item_id: string; label?: string }[];
      dropped?: { source_key: string; item_id: string; label?: string }[];
      missing?: { source_key: string; item_id: string; label?: string }[];
    };
    html?: string;
  }[];
  final?: { finished_at?: string; version?: number; summary?: string };
}

export interface CvGeneratePreviewOut {
  stage: string;
  pct: number;
  iteration: number;
  html: string;
  trace?: CvPolishTrace;
}

export interface CvGenerateStatus {
  job_id: string;
  status: "queued" | "running" | "succeeded" | "failed" | "cancelled";
  progress: number;
  stage: string | null;
  error: string | null;
  result: CvGenerateResult | null;
  created_at: string;
  finished_at: string | null;
}

export interface CvWorkingContent {
  blocks?: CvBlock[];
  overrides?: Record<string, Record<string, string>>;
  [key: string]: unknown;
}

export type CvAreaId = "main" | "sidebar";

export interface CvBlock {
  kind: string;
  props?: Record<string, unknown>;
  area?: CvAreaId;
  column?: CvAreaId;
}

export interface CvDocumentOut {
  id: string;
  title: string;
  kind: "resume" | "cover_letter";
  target_posting_id: string | null;
  template_id: string | null;
  photo_document_id: string | null;
  language: string;
  page_size: "a4" | "letter";
  max_pages: number;
  status: "draft" | "final" | "archived";
  working_content: CvWorkingContent;
  context: Partial<CvContextSelection>;
  source_document_id: string | null;
  created_at: string;
  updated_at: string;
  latest_version: number | null;
}

export interface CvDocumentCreate {
  title: string;
  kind?: CvDocumentOut["kind"];
  language?: string;
  page_size?: CvDocumentOut["page_size"];
  max_pages?: number;
  target_posting_id?: string | null;
  template_id?: string | null;
  photo_document_id?: string | null;
  source_document_id?: string | null;
  context?: CvContextSelection;
}

export interface CvDocumentUpdate {
  title?: string;
  status?: CvDocumentOut["status"];
  language?: string;
  page_size?: CvDocumentOut["page_size"];
  max_pages?: number;
  target_posting_id?: string | null;
  template_id?: string | null;
  photo_document_id?: string | null;
  working_content?: CvWorkingContent;
  context?: CvContextSelection;
}

export interface CvContextResolutionRef {
  source_key: string;
  item_id: string;
  label?: string;
  updated_at?: string;
}

export interface CvContextItemInfo {
  item_id: string;
  label: string;
  detail: string;
}

export interface CvContextSourceOut {
  key: string;
  label: string;
  description: string;
  items: CvContextItemInfo[];
}

export interface CvContextSourcesOut {
  sources: CvContextSourceOut[];
}

export interface CvRenderMetrics {
  estimated_lines?: number;
  lines_per_page?: number;
  estimated_pages?: number;
  empty_blocks?: string[];
  truncated?: number;
  overflow?: boolean;
  max_pages?: number;
  page_size?: string;
  [key: string]: unknown;
}

export interface CvPreviewOut {
  html: string;
  metrics: CvRenderMetrics;
  blocks: CvBlock[];
  resolution: {
    snapshot: Record<string, unknown>;
    snapshot_index: Record<string, string[]>;
    items: CvContextResolutionRef[];
    resolved_at: string;
  };
}

export interface CvVersionOut {
  id: string;
  version: number;
  content: Record<string, unknown>;
  context_resolution: {
    resolved_at?: string;
    items?: CvContextResolutionRef[];
  };
  content_hash: string;
  created_by: string;
  created_at: string;
}

export interface CvCompileOut {
  version: CvVersionOut;
  html: string;
  metrics: CvRenderMetrics;
}

export interface CvContextStatusOut {
  has_baseline: boolean;
  stale: boolean;
  changed: CvContextResolutionRef[];
  added: CvContextResolutionRef[];
  removed: CvContextResolutionRef[];
}

export interface CvLintCheck {
  id: string;
  level: "pass" | "info" | "warn" | "fail";
  message: string;
}

export interface CvLintReport {
  score: number;
  passed: boolean;
  checks: CvLintCheck[];
  metrics: CvRenderMetrics;
  resolved_items: number;
}

export interface CvProposal {
  ref?: CvContextRef;
  field?: string;
  text: string;
  rationale: string;
  evidence_refs: CvContextRef[];
}

export interface VerifiedProposal {
  proposal: CvProposal;
  verified: boolean;
}

export interface CoverageEntry {
  skill_key: string;
  label: string;
  priority: string;
  user_level: number | null;
}

export interface TailorCoverage {
  covered: CoverageEntry[];
  missing: CoverageEntry[];
}

export interface SectionGap {
  source_key: string;
  label: string;
  message: string;
  candidate_count: number;
}

export interface CvSuggestionOut {
  action: string;
  notes: string;
  proposals: VerifiedProposal[];
  coverage: TailorCoverage | null;
  gaps: SectionGap[];
}

export interface LetterParagraph {
  text: string;
  evidence_refs: CvContextRef[];
}

export interface CoverLetterDraft {
  subject: string;
  salutation: string;
  paragraphs: LetterParagraph[];
  closing: string;
}

export interface VerifiedParagraph {
  text: string;
  evidence_refs: CvContextRef[];
  verified: boolean;
}

export interface CoverLetterSuggestionOut {
  action: "cover_letter";
  notes: string;
  draft: CoverLetterDraft;
  paragraphs: VerifiedParagraph[];
}

export interface BriefSkill {
  skill_key: string;
  label: string;
  required_level: number;
  priority: string;
  evidence_quote: string;
  user_level: number | null;
}

export interface BriefFitDimension {
  dimension: string;
  score: number;
  detail: string;
}

export interface CoverLetterBriefOut {
  posting_id: string;
  posting_title: string;
  org: string;
  location: string;
  extract_ready: boolean;
  must_have: BriefSkill[];
  nice_to_have: BriefSkill[];
  responsibilities: string[];
  fit: {
    score: number | null;
    estimate: boolean;
    dimensions: BriefFitDimension[];
  };
  coverage: TailorCoverage;
  goal: string;
  evidence_items: number;
}

export interface CoverLetterCreate {
  posting_id: string;
  title?: string;
  base_cv_id?: string;
  language?: string;
}
/** One audited LLM call of a CV run (plan 65.3). */
export interface CvRunCall {
  id: string;
  task: string;
  stage?: string | null;
  status: string;
  provider: string;
  model: string;
  prompt_version?: string | null;
  tokens_in?: number | null;
  tokens_out?: number | null;
  latency_ms?: number | null;
}

/** One run of `cv_generate` / `cv_polish` for a CV (plan 65.3). */
export interface CvRunOut {
  job_id: string;
  job_type: string;
  status: string;
  stage?: string | null;
  error?: string | null;
  created_at: string;
  finished_at?: string | null;
  outcome?: string | null;
  resumed_from?: string | null;
  final_version?: number | null;
  stages?: { node: string; at: string; note?: string }[];
  iterations?: {
    n: number;
    summary?: string;
    ops?: { op: string; ok: boolean; detail?: string }[];
    coverage?: {
      included?: { source_key: string; item_id: string; label?: string }[];
      dropped?: { source_key: string; item_id: string; label?: string }[];
      missing?: { source_key: string; item_id: string; label?: string }[];
    };
  }[];
  llm_calls: CvRunCall[];
  aggregate: {
    calls: number;
    tokens_in: number;
    tokens_out: number;
    latency_ms_sum: number;
    by_task: Record<
      string,
      { calls: number; tokens_in: number; tokens_out: number; latency_ms_sum: number }
    >;
  };
}
